src/syntax/ppx_sedlex.ml

```diff
@@ -253,24 +264,228 @@ let gen_recflag auto =
+              let guard =
+                if Hashtbl.find dup_case (curr_state, curr_node, prev_state)
+                then None
+                else Some (gen_cset ~loc "__sedlex_code" char_set)
+              in
+              let rhs =
+                let call_rest =
+                  [%expr
+                    __sedlex_aux (__sedlex_pos - 1) [%e eint ~loc prev_state]
+                      [%e eint ~loc prev_node] __sedlex_rest]
+                in
+                List.fold_left gen_action call_rest
+                  (List.map find_offset_idx actions |> List.sort_uniq compare)
+              in
+              case ~lhs ~guard ~rhs)
+            trans
+        in
+        let final_cases =
+          List.map
+            (fun { Sedlex.curr_node; actions } ->
+              let lhs = pint ~loc curr_node in
+              let rhs =
+                List.fold_left gen_action [%expr ()]
+                  (List.map find_offset_idx actions |> List.sort_uniq compare)
+              in
+              case ~lhs ~guard:None ~rhs)
+            finals
+        in
+        value_binding ~loc
+          ~pat:[%pat? __sedlex_aux]
+          ~expr:
+            [%expr
+              fun __sedlex_pos __sedlex_curr_state __sedlex_curr_node ->
+                function
+                | [] ->
+                    [%e
+                      pexp_match ~loc [%expr __sedlex_curr_node]
+                        (final_cases @ [unreachable_case])]
+                | __sedlex_prev_state :: __sedlex_rest ->
+                    let __sedlex_code =
+                      Sedlexing.lexeme_code [%e evar ~loc lexbuf]
+                        (__sedlex_pos - 1)
+                    in
+                    [%e
+                      pexp_match ~loc
+                        [%expr
+                          __sedlex_curr_state,
+                            __sedlex_curr_node,
+                            __sedlex_prev_state]
+                        (trans_cases @ [unreachable_case])]]
+      in
+      [
+        value_binding ~loc
+          ~pat:(pvar ~loc (trace_fun i))
+          ~expr:
+            [%expr
+              fun [%p pvar ~loc lexbuf] __sedlex_path ->
+                [%e
+                  pexp_let ~loc Nonrecursive [offset_array]
+                  @@ pexp_let ~loc Recursive [aux_fun]
+                  @@ [%expr
+                       (match __sedlex_path with
+                         | __sedlex_curr_state :: __sedlex_rest ->
+                             __sedlex_aux
+                               (Sedlexing.lexeme_length [%e evar ~loc lexbuf])
+                               __sedlex_curr_state [%e eint ~loc initial]
+                               __sedlex_rest
+                         | _ -> assert false);
+                       __sedlex_offsets]]];
+      ]
+
+let gen_aliases lexbuf offsets i e = function
+  | [] -> e
+  | aliases ->
+      let loc = default_loc in
+      let _, action_offsets = Option.get offsets.(i) in
```

Comment:
instead of matching on `aliases` above, we could directly match on `offsets.(i)`
```ocaml
let gen_aliases lexbuf offsets i e aliases =
  match offsets.(i) with
  | None -> e
  | Some (_, action_offsets) ->
     let loc = default_loc in
     pexp_let ~loc Nonrecursive
        [
          value_binding ~loc
            ~pat:[%pat? __sedlex_offsets]
            ~expr:
              (appfun (trace_fun i) [evar ~loc lexbuf; [%expr __sedlex_path]]);
        ]
      @@ pexp_let ~loc Nonrecursive
           (List.map
             (fun { txt = alias; loc } ->
                let start = Hashtbl.find action_offsets (alias, true) in
                let stop = Hashtbl.find action_offsets (alias, false) in
               value_binding ~loc ~pat:(pvar ~loc alias)
                  ~expr:
                    [%expr
                      __sedlex_offsets.([%e eint ~loc start]),
                        __sedlex_offsets.([%e eint ~loc stop])
                        - __sedlex_offsets.([%e eint ~loc start])])
              aliases)
      @@ e
```